### PR TITLE
Refactor code base to reduce memory requirement for building libcugraph

### PR DESCRIPTION
**Pull request by naimnv**
Refactor code base to reduce memory requirement for building libcugraph.

- split all sg.cu and mg.cu files into three files for three combination of (vertex_t, edge_t ), i.e, (32 bit, 32bit), (32 bit, 64bit), (64 bit, 64bit)

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
+    src/link_prediction/jaccard_sg64.cu
+    src/link_prediction/jaccard_sg32.cu
+    src/link_prediction/jaccard_sg3264.cu
+    src/link_prediction/sorensen_sg64.cu
+    src/link_prediction/sorensen_sg32.cu
+    src/link_prediction/sorensen_sg3264.cu
+    src/link_prediction/overlap_sg64.cu
+    src/link_prediction/overlap_sg32.cu
+    src/link_prediction/overlap_sg3264.cu
     src/link_prediction/cosine_sg.cu
```
Comment (seunghwak) [inline cpp/CMakeLists.txt:222]: Why not split cosine_sg.cu?

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
     src/detail/collect_local_vertex_values.cu
```
Comment (seunghwak) [inline cpp/CMakeLists.txt:174]: This file also explicitly instantiates based on vertex_t, value_t, and multi_gpu, should I create sg32, sg64, mg32, and mg64?

File: cpp/CMakeLists.txt [hunk 3]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
     src/detail/groupby_and_count.cu
     src/detail/collect_comm_wrapper.cu
```
Comment (seunghwak) [inline cpp/CMakeLists.txt:176]: Similarly, this file is MG only but explicitly instantiates based on vertex_t, edge_t, and weight_t. May create mg32, mg3264, and mg64?

File: cpp/CMakeLists.txt [hunk 4]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
+    src/community/detail/maximal_independent_moves_sg3264.cu
+    src/community/detail/maximal_independent_moves_mg64.cu
+    src/community/detail/maximal_independent_moves_mg32.cu
+    src/community/detail/maximal_independent_moves_mg3264.cu
     src/detail/utility_wrappers.cu
-    src/structure/graph_view_mg.cu
+    src/structure/graph_view_mg64.cu
+    src/structure/graph_view_mg32.cu
+    src/structure/graph_view_mg3264.cu
     src/structure/remove_self_loops.cu
```
Comment (seunghwak) [inline cpp/CMakeLists.txt:208]: This function is SG only but explicitly instantiates based on vertex_t and edge_t, no need to split? (create sg32, sg3264, and sg64?)

File: cpp/CMakeLists.txt [hunk 5]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
     src/structure/remove_multi_edges.cu
```
Comment (seunghwak) [inline cpp/CMakeLists.txt:209]: This function is SG only but explicitly instantiates based on vertex_t and edge_t, no need to split?

File: cpp/CMakeLists.txt [hunk 6]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
     src/utilities/path_retrieval.cu
```
Comment (seunghwak) [inline cpp/CMakeLists.txt:210]: This function is SG only but explicitly instantiates based on vertex_t, no need to split?

File: cpp/CMakeLists.txt [hunk 7]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
+    src/link_prediction/jaccard_mg64.cu
+    src/link_prediction/jaccard_mg32.cu
+    src/link_prediction/jaccard_mg3264.cu
+    src/link_prediction/sorensen_mg64.cu
+    src/link_prediction/sorensen_mg32.cu
+    src/link_prediction/sorensen_mg3264.cu
+    src/link_prediction/overlap_mg64.cu
+    src/link_prediction/overlap_mg32.cu
+    src/link_prediction/overlap_mg3264.cu
     src/link_prediction/cosine_mg.cu
```
Comment (seunghwak) [inline cpp/CMakeLists.txt:232]: Why not split cosine_mg?

**Comment by bdice** [comment]
Would a naming scheme like `..._sg_v32_e64.cu` be acceptable?

I think separators between the numerical values and a one-character label like `v` or `e` would be more readable and convey more helpful information. If you wanted to split by weights float/double later, you could suffix with `_wf` or `_wd`.

File: cpp/CMakeLists.txt (hunk 3 above)
Comment (naimnv) [reply cpp/CMakeLists.txt:176]: Already working on it

File: cpp/CMakeLists.txt [hunk 8]
```diff
@@ -168,135 +168,335 @@ endif()
...
-    src/community/detail/maximal_independent_moves_mg.cu
-    src/detail/utility_wrappers.cu
-    src/structure/graph_view_mg.cu
-    src/structure/remove_self_loops.cu
-    src/structure/remove_multi_edges.cu
-    src/utilities/path_retrieval.cu
+    src/utilities/shuffle_vertices_mg.cu
+    src/detail/permute_range_32.cu
+    src/detail/permute_range_64.cu
+    src/utilities/shuffle_vertex_pairs_mg.cu
```
Comment (seunghwak) [inline cpp/CMakeLists.txt:174]: Don't we need to further split this file? This is one of the most problematic files.

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (naimnv) [reply cpp/CMakeLists.txt]: Done.

File: cpp/CMakeLists.txt (hunk 3 above)
Comment (naimnv) [reply cpp/CMakeLists.txt]: Done.

File: cpp/CMakeLists.txt (hunk 4 above)
Comment (naimnv) [reply cpp/CMakeLists.txt]: Done.

File: cpp/CMakeLists.txt (hunk 6 above)
Comment (naimnv) [reply cpp/CMakeLists.txt]: Done.

File: cpp/CMakeLists.txt [hunk 9]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
+    src/community/egonet_sg64.cu
+    src/community/egonet_sg32.cu
+    src/community/egonet_sg3264.cu
+    src/community/egonet_mg64.cu
+    src/community/egonet_mg32.cu
+    src/community/egonet_mg3264.cu
+    src/community/k_truss_sg64.cu
+    src/community/k_truss_sg32.cu
+    src/community/k_truss_sg3264.cu
     src/sampling/random_walks.cu
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/CMakeLists.txt]: Done.

File: cpp/CMakeLists.txt (hunk 7 above)
Comment (naimnv) [reply cpp/CMakeLists.txt]: Done.

File: cpp/CMakeLists.txt [hunk 10]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
+    src/sampling/detail/prepare_next_frontier_sg32.cu
+    src/sampling/detail/prepare_next_frontier_mg64.cu
+    src/sampling/detail/prepare_next_frontier_mg32.cu
+    src/sampling/detail/gather_one_hop_edgelist_sg64.cu
+    src/sampling/detail/gather_one_hop_edgelist_sg32.cu
+    src/sampling/detail/gather_one_hop_edgelist_sg3264.cu
+    src/sampling/detail/gather_one_hop_edgelist_mg64.cu
+    src/sampling/detail/gather_one_hop_edgelist_mg32.cu
+    src/sampling/detail/gather_one_hop_edgelist_mg3264.cu
     src/sampling/detail/remove_visited_vertices_from_frontier.cu
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/CMakeLists.txt]: Done.

File: cpp/CMakeLists.txt [hunk 11]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
+    src/cores/core_number_mg3264.cu
+    src/cores/k_core_sg64.cu
+    src/cores/k_core_sg32.cu
+    src/cores/k_core_sg3264.cu
+    src/cores/k_core_mg64.cu
+    src/cores/k_core_mg32.cu
+    src/cores/k_core_mg3264.cu
     src/components/legacy/connectivity.cu
     src/generators/generate_rmat_edgelist.cu
     src/generators/generate_bipartite_rmat_edgelist.cu
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/CMakeLists.txt]: Done.

File: cpp/CMakeLists.txt [hunk 12]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
     src/generators/generator_tools.cu
     src/generators/simple_generators.cu
     src/generators/erdos_renyi_generator.cu
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/CMakeLists.txt]: Done.

File: cpp/CMakeLists.txt [hunk 13]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
+    src/community/approx_weighted_matching_mg32.cu
+    src/community/approx_weighted_matching_mg3264.cu
+    src/traversal/k_hop_nbrs_sg64.cu
+    src/traversal/k_hop_nbrs_sg32.cu
+    src/traversal/k_hop_nbrs_sg3264.cu
+    src/traversal/k_hop_nbrs_mg64.cu
+    src/traversal/k_hop_nbrs_mg32.cu
+    src/traversal/k_hop_nbrs_mg3264.cu
     src/mtmg/vertex_result.cu
     src/mtmg/vertex_pairs_result.cu
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/CMakeLists.txt]: Done.

File: cpp/CMakeLists.txt [hunk 14]
```diff
@@ -174,127 +174,312 @@ set(CUGRAPH_SOURCES
...
+    src/sampling/detail/sample_edges_sg32.cu
+    src/sampling/detail/sample_edges_sg3264.cu
+    src/sampling/detail/sample_edges_mg64.cu
+    src/sampling/detail/sample_edges_mg32.cu
+    src/sampling/detail/sample_edges_mg3264.cu
+    src/sampling/detail/shuffle_and_organize_output_mg64.cu
+    src/sampling/detail/shuffle_and_organize_output_mg32.cu
+    src/sampling/detail/shuffle_and_organize_output_mg3264.cu
     src/sampling/uniform_neighbor_sampling_mg.cpp
     src/sampling/uniform_neighbor_sampling_sg.cpp
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/CMakeLists.txt]: Done.

**Review by seunghwak — APPROVED** [review]
LGTM!!!

**Review by ChuckHastings — APPROVED** [review]
Ugly but effective.

**Review by jnke2016 — APPROVED** [review]

**Pull request merged**